src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,678 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using System.Text;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    /// <summary>
+    /// A media type value.
+    /// </summary>
+    public struct MediaType
+    {
+        private static readonly StringSegment QualityParameter = new StringSegment("q");
+
+        private readonly string _mediaType;
+        private readonly int _offset;
+        private readonly int? _length;
+        private Enumerator _parameterEnumerator;
+
+        /// <summary>
+        /// Initializes a <see cref="MediaType"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="string"/> with the media type.</param>
+        public MediaType(string mediaType)
+            : this(mediaType, 0, mediaType.Length)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a <see cref="MediaType"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="StringSegment"/> with the media type.</param>
+        public MediaType(StringSegment mediaType)
+            : this(mediaType.Buffer, mediaType.Offset, mediaType.Length)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a <see cref="MediaTypeParser"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="string"/> with the media type.</param>
+        /// <param name="offset">The offset in the <paramref name="mediaType"/> where the parsing starts.</param>
+        /// <param name="length">The of the media type to parse if provided.</param>
+        public MediaType(string mediaType, int offset, int? length)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            if (offset < 0 || offset >= mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (length != null && offset + length > mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            _mediaType = mediaType;
+            _offset = offset;
+            _length = length;
+            _parameterEnumerator = default(Enumerator);
+            Type = new StringSegment();
+            SubType = new StringSegment();
+
+            var enumerator = GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                return;
+            }
+            Type = enumerator.Current.Value;
+
+            if (!enumerator.MoveNext())
+            {
+                return;
+            }
+            SubType = enumerator.Current.Value;
```

Comment:
Would it be clear if instead of this being an "Enumerator" was a "State machine"? in the end is what it is. I can make it MediaTypeParsingState instead of the Enumerator, if that clarifies things.